pkg/apis/resource/validation/validation.go

```diff
@@ -729,3 +738,77 @@ func truncateIfTooLong(str string, maxLen int) string {
 	remaining := maxLen - len(ellipsis)
 	return str[0:(remaining+1)/2] + ellipsis + str[len(str)-remaining/2:]
 }
+
+func validateDeviceStatus(device resource.AllocatedDeviceStatus, fldPath *field.Path, allocatedDevices sets.Set[structured.DeviceID]) field.ErrorList {
+	var allErrs field.ErrorList
+	allErrs = append(allErrs, validateDriverName(device.Driver, fldPath.Child("driver"))...)
+	allErrs = append(allErrs, validatePoolName(device.Pool, fldPath.Child("pool"))...)
+	allErrs = append(allErrs, validateDeviceName(device.Device, fldPath.Child("device"))...)
+	deviceID := structured.MakeDeviceID(device.Driver, device.Pool, device.Device)
+	if !allocatedDevices.Has(deviceID) {
+		allErrs = append(allErrs, field.Invalid(fldPath, deviceID, "must be an allocated device in the claim"))
+	}
+	allErrs = append(allErrs, metav1validation.ValidateConditions(device.Conditions, fldPath.Child("conditions"))...)
+	if len(device.Data.Raw) > 0 { // Data is an optional field.
+		allErrs = append(allErrs, validateRawExtension(device.Data, fldPath.Child("data"), false)...)
+	}
+	allErrs = append(allErrs, validateNetworkDeviceData(device.NetworkData, fldPath.Child("networkData"))...)
+	return allErrs
+}
+
+// validateRawExtension validates RawExtension as in https://github.com/kubernetes/kubernetes/pull/125549/
+func validateRawExtension(rawExtension runtime.RawExtension, fldPath *field.Path, stored bool) field.ErrorList {
+	var allErrs field.ErrorList
+	var v any
+	if len(rawExtension.Raw) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath, ""))
+	} else if !stored && len(rawExtension.Raw) > resource.OpaqueParametersMaxLength {
+		// Don't even bother with parsing when too large.
+		// Only applies on create. Existing parameters are grand-fathered in
+		// because the limit was introduced in 1.32. This also means that it
+		// can be changed in the future.
+		allErrs = append(allErrs, field.TooLong(fldPath, "" /* unused */, resource.OpaqueParametersMaxLength))
+	} else if err := json.Unmarshal(rawExtension.Raw, &v); err != nil {
+		allErrs = append(allErrs, field.Invalid(fldPath, "<value omitted>", fmt.Sprintf("error parsing data as JSON: %v", err.Error())))
+	} else if v == nil {
+		allErrs = append(allErrs, field.Required(fldPath, ""))
+	} else if _, isObject := v.(map[string]any); !isObject {
+		allErrs = append(allErrs, field.Invalid(fldPath, "<value omitted>", "parameters must be a valid JSON object"))
```

Comment:
This is making the validation of Raw extension for general purpose but is keeping the references to the original implementation, that is `config.Parameters.Raw` and we are also calling this from `device.Data` .

This can be confusing when logging  , maybe we should use WET here instead of DRY https://en.wikipedia.org/wiki/Don%27t_repeat_yourself , we are only writing it twice literally

Reply:
> // Only applies on create. Existing parameters are grand-fathered in
		// because the limit was introduced in 1.32. This also means that it
		// can be changed in the future.


and 

> allErrs = append(allErrs, field.Invalid(fldPath, "<value omitted>", "parameters must be a valid JSON object"))